corona-warn-app/cwa-app-ios
# Story/10291 most relevant dcc

## Description
This PR changes the criteria for most relevant dcc

## Acceptance Criteria
1. When determining the most relevant DCC, a booster has priority over another VC without the 14-day period
2. When determining the most relevant DCC, a VC that indicates vaccination after a recovery has priority over another VC without the 14-day period
3. When determining the most relevant DCC, a VC with an mRNA vaccine after Janssen is subject to the 14-day period (current default behavior)

## Test cases
Case 1:
Booster after Moderna, Biontech, Astra (2/2) -> gets priority 
Booster after J&J (1/1) -> gets priority 

Info texts shows complete vaccination

Case 2:
Booster with Moderna, Biontech, Astra (2/2) after Recovery Vaccination (1/1) -> gets priority after 14 days 

Info texts shows complete vaccination

Case 3:
Booster with Moderna, Biontech, Astra (2/2) after J&J (1/1) -> gets priority after 14 days 

Info texts shows complete vaccination

Case 4:
Vaccination with Moderna, Biontech, Astra (1/1) after recovery -> gets priority 

Info texts shows complete vaccination

Case 5:
Vaccination with J&J (1/1) after recovery -> get priority after 14 days 

Info texts shows complete vaccination after 14 days

## Steps for testing
1. Scan a recovery certificate < 180 days old:
cwa rc gen --pii 42 -s r.0.fr='2021-08-01' r.0.df='2021-10-10' -e wru. This will have highest prio as it's the only one.

2. Scan a J&J vaccine < 14 days old:
cwa vc gen --pii 42 -s v.0.dn=1 v.0.sd=1 v.0.mp='EU/1/20/1525' v.0.dt='2021-11-04' -e wru. The recovery certificate should still have highest prio as this vaccine is < 14 days old. Vaccination Status info box should say you have received all planned vaccinations but protection won't be complete for x days.

3. Scan a J&J vaccine > 14 days old:
cwa vc gen --pii 42 -s v.0.dn=1 v.0.sd=1 v.0.mp='EU/1/20/1525' v.0.dt='2021-10-04' -e wru. This vaccine will now get the highest prio as the 14 days have passed. Vaccination Status info box should say you have received all planned vaccinations and protection is complete.

Remove the two J&J vaccines and leave the recovery certificate in place.

4. Scan a BioNTech vaccine with dose 1/1 that's < 14 days old:
cwa vc gen --pii 42 -s v.0.dn=1 v.0.sd=1 v.0.mp='EU/1/20/1528' v.0.dt='2021-11-04' -e wru. This vaccine should get highest priority right away. Vaccination Status info box should say you have received all planned vaccinations and protection is complete.

5. Scan a BioNTech vaccine with dose 1/1 that's > 14 days old:
cwa vc gen --pii 42 -s v.0.dn=1 v.0.sd=1 v.0.mp='EU/1/20/1528' v.0.dt='2021-10-04' -e wru. The most recent BioNTech vaccine (the one < 14 days from the previous point) should still have highest prio. Vaccination Status info box should say you have received all planned vaccinations and protection is complete.

Remove the two BioNTech vaccines.

6. Scan a J&J vaccine > 14 days old:
cwa vc gen --pii 42 -s v.0.dn=1 v.0.sd=1 v.0.mp='EU/1/20/1525' v.0.dt='2021-10-04' -e wru. This will have the highest prio (already covered at point 3).

7. Scan a BioNTech vaccine with dose 2/2 that's < 14 days old:
cwa vc gen --pii 42 -s v.0.dn=2 v.0.sd=2 v.0.mp='EU/1/20/1528' v.0.dt='2021-11-04' -e wru. The J&J vaccine still has the highest priority as this vaccine is < 14 days old. The Vaccination Status info box should say you have received all planned vaccinations.

8. Scan a BioNTech vaccine with dose 2/2 that's > 14 days old:
cwa vc gen --pii 42 -s v.0.dn=2 v.0.sd=2 v.0.mp='EU/1/20/1528' v.0.dt='2021-10-04' -e wru. This vaccine now gets highest priority. The Vaccination Status info box should say you have received all planned vaccinations and protection is complete.

Delete all vaccination certificates and leave the recovery certificate in place.

9. Scan a BioNTech certificate with dose 1/1 that's > 14 days old:
cwa vc gen --pii 42 -s v.0.dn=1 v.0.sd=1 v.0.mp='EU/1/20/1528' v.0.dt='2021-10-04' -e wru. This vaccine will have highest priority. The Vaccination Status info box should say you have recieved all planned vaccinations and protection is complete.

10. Scan a BioNTech certificate with dose 2/2 that's < 14 days old:
cwa vc gen --pii 42 -s v.0.dn=2 v.0.sd=2 v.0.mp='EU/1/20/1528' v.0.dt='2021-11-04' -e wru. The previous BioNTech vaccine with dose 1/1 will still have highest priority as this one is < 14 days old. The Vaccination Status info box should say you have received all planned vaccinations.

11. Scan a BioNTech certificate with dose 2/2 that's > 14 days old:
cwa vc gen --pii 42 -s v.0.dn=2 v.0.sd=2 v.0.mp='EU/1/20/1528' v.0.dt='2021-10-04' -e wru. This vaccine will now get highest priority. The Vaccination Status info box should say you have received all planned vaccinations and protection is complete.

Remove all certificates.

12. Scan a BioNTech certificate with dose 1/2 that's < 14 days old:
cwa vc gen --pii 42 -s v.0.dn=1 v.0.sd=2 v.0.mp='EU/1/20/1528' v.0.dt='2021-11-04' -e wru. This vaccine will have highest prio as it's the only certificate scanned. The Vaccination Status info box should say you have not yet received all planned vaccinations. Your protection is not complete yet.

13. Scan a BioNTech certificate with dose 1/2 that's > 14 days old:
cwa vc gen --pii 42 -s v.0.dn=1 v.0.sd=2 v.0.mp='EU/1/20/1528' v.0.dt='2021-10-04' -e wru. The previous vaccine that's < 14 days old will still have highest priority because for incomplete doses we take the priority depending on the most recent date. The Vaccination Status info box should say you have not yet received all planned vaccinations. Your protection is not complete yet.

14. Scan a BioNTech certificate with dose 2/2 that's < 14 days old:
cwa vc gen --pii 42 -s v.0.dn=2 v.0.sd=2 v.0.mp='EU/1/20/1528' v.0.dt='2021-11-04' -e wru. This vaccination will get highest priority as it's a complete vaccination. The Vaccination Status info box should say you have received all planned vaccinations but protection won't be complete for x days.

Remove the previous vaccine with dose 2/2 and < 14 days old.

15. Scan a BioNTech certificate with dose 2/2 that's > 14 days old:
cwa vc gen --pii 42 -s v.0.dn=2 v.0.sd=2 v.0.mp='EU/1/20/1528' v.0.dt='2021-10-04' -e wru. This vaccine will get highest priority. The Vaccination Status info box should say you have recieved all planned vaccinations and protection is complete.

16. Scan a BioNTech certificate with dose 3/3 that's < 14 days old:
cwa vc gen --pii 42 -s v.0.dn=3 v.0.sd=3 v.0.mp='EU/1/20/1528' v.0.dt='2021-11-04' -e wru. This vaccine will get highest priority right away. The Vaccination Status info box should say you have received all planned vaccinations and protection is complete.

Remove all certificates.

17. Scan a J&J with dose 1/1 that's > 14 days old:
cwa vc gen --pii 42 -s v.0.dn=1 v.0.sd=1 v.0.mp='EU/1/20/1525' v.0.dt='2021-10-04' -e wru. This vaccine will get highest priority as it's the only certificate. The Vaccination Status info box should say you have received all planned vaccinations and protection is complete.

18. Scan a J&J with dose 2/2 that's < 14 days old:
cwa vc gen --pii 42 -s v.0.dn=2 v.0.sd=2 v.0.mp='EU/1/20/1525' v.0.dt='2021-11-04' -e wru. This vaccine will get highest priority right away as it's considered a booster. The Vaccination Status info box should say you have received all planned vaccinations.

Credits for creating the steps: @SamuraiKek 

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10291
